### ISSUE TITLE
Encoding methods

### ISSUE THREAD

**Issue by slucas03** [fangwei123456/spikingjelly#182]
I have recently started to work with SpikingJelly and I have queries about the kind of encoding methods that SpikingJelly is able to work with. In other words, I have noticed that the encoding methods that are explained in the documentation ([https://github.com/fangwei123456/spikingjelly/blob/master/docs/source/clock_driven_en/2_encoding.rst](url)) are mostly based on rate coding, such as, Poisson or weighted phase. In some papers is mentioned that the type of enconding method influences in the training of spiking neural network. So, I do not know whether SpikingJelly is prepared to work with timing coding methods or it must be used only with rate coding.

Thanks in advance.

**Comment by fangwei123456** [comment]
Hi, the decoding method is independent with the SNN. So, you can use any encoding method for SNN built by SJ.
We have defined a temporal encoder, the latency encoder (Time-To-First-Spike, TTFS): https://spikingjelly.readthedocs.io/zh_CN/latest/spikingjelly.clock_driven.encoding.html#latencyencoder-init-en

You can use it to train a SNN on CIFAR-10 and get about 80+% accuracy.

**Comment by slucas03** [comment]
Thank you very much for your quick answer! I am going to try right now :)

**Comment by fangwei123456** [comment]
>is there any possibility to define the exact position of the spike in the neuron's time window

Yes, you can use `spike * t` to get the firing time:
```python
import torch
import torch.nn as nn
import torch.nn.functional as F
T = 8
spike = (torch.rand([T]) > 0.5).float()
t = torch.arange(T)

print(f'spike = {spike}')
t_f = spike * t
print(f't_f = {t_f}')
mask = spike == 1
print(f't_f[mask] = {t_f[mask]}')
```

The outputs are 
```
spike = tensor([1., 1., 0., 0., 0., 0., 1., 1.])
t_f = tensor([0., 1., 0., 0., 0., 0., 6., 7.])
t_f[mask] = tensor([0., 1., 6., 7.])
```

You can get another useage at https://github.com/fangwei123456/spikingjelly/blob/master/publications.md .

For example:

Deep Q-learning: https://github.com/AptX395/Deep-Spiking-Q-Networks

Regression: https://github.com/urancon/StereoSpike

**Comment by slucas03** [comment]
Thank you for your comment. Very nice and helpful info!!

**Comment by bhzhang95** [comment]
> Hi, the encoding method is independent with the SNN. So, you can use any encoding method for SNN built by SJ. We have defined a temporal encoder, the latency encoder (Time-To-First-Spike, TTFS): https://spikingjelly.readthedocs.io/zh_CN/latest/spikingjelly.clock_driven.encoding.html#latencyencoder-init-en
> 
> You can use it to train a SNN on CIFAR-10 and get about 80+% accuracy.

Where can I find the example of time-to-first-spike coding SNN on CIFAR-10 dataset?

**Comment by fangwei123456** [comment]
> > Hi, the encoding method is independent with the SNN. So, you can use any encoding method for SNN built by SJ. We have defined a temporal encoder, the latency encoder (Time-To-First-Spike, TTFS): https://spikingjelly.readthedocs.io/zh_CN/latest/spikingjelly.clock_driven.encoding.html#latencyencoder-init-en
> > You can use it to train a SNN on CIFAR-10 and get about 80+% accuracy.
> 
> Where can I find the example of time-to-first-spike coding SNN on CIFAR-10 dataset?

You can try to use this encoder to generate input spikes, and train a SNN based on SpikingJelly. For example, you can use the PLIF net for CIFAR10:

https://github.com/fangwei123456/spikingjelly/blob/0bdbe41480656b0e247575b338e98424fb9bfabd/spikingjelly/activation_based/model/parametric_lif_net.py#L50

You can also try it on FashionMNIST. The network can be obrained from the tutorial. The only difference is that you should modify the input.

**Comment by bhzhang95** [comment]
> > > Hi, the encoding method is independent with the SNN. So, you can use any encoding method for SNN built by SJ. We have defined a temporal encoder, the latency encoder (Time-To-First-Spike, TTFS): https://spikingjelly.readthedocs.io/zh_CN/latest/spikingjelly.clock_driven.encoding.html#latencyencoder-init-en
> > > You can use it to train a SNN on CIFAR-10 and get about 80+% accuracy.
> > 
> > 
> > Where can I find the example of time-to-first-spike coding SNN on CIFAR-10 dataset?
> 
> You can try to use this encoder to generate input spikes, and train a SNN based on SpikingJelly. For example, you can use the PLIF net for CIFAR10:
> 
> https://github.com/fangwei123456/spikingjelly/blob/0bdbe41480656b0e247575b338e98424fb9bfabd/spikingjelly/activation_based/model/parametric_lif_net.py#L50
> 
> You can also try it on FashionMNIST. The network can be obrained from the tutorial. The only difference is that you should modify the input.

When using the time-to-first-spike coding, how can I define the loss function? I've tried to train a mlp to classify MNIST, however, the loss cannot converge

**Comment by fangwei123456** [comment]
You can refer to those papers that using the TTFS-based SNNs.

**Comment by AllenYolk** [comment]
SpikingJelly 不绑定编码方式——脉冲编码与 SNN 本身是正交的，框架只提供编码器与神经元的组合能力。

`spikingjelly.activation_based.encoding` 内置的编码器：

- `PoissonEncoder`：泊松 / 频率编码（Stateless）
- `LatencyEncoder`：latency / TTFS 编码（Stateful）
- `PeriodicEncoder`：周期脉冲
- `WeightedPhaseEncoder`：加权相位
- `PopSpikeEncoderDeterministic` / `PopSpikeEncoderRandom`：群体编码
- `PopEncoder`：population rate encoding

选择编码方式的简单准则：
- 性能 / 通用性首选 **频率编码（Poisson / 泊松）**
- 需要稀疏、能量友好时考虑 **延迟 / TTFS**（训练难度更高）
- 需要 spike timing 精确控制时考虑 **TTFS** 或自实现神经元

教程：https://spikingjelly.readthedocs.io/en/latest/activation_based/4_encoding.html

**Comment by AllenYolk** [comment]
SpikingJelly does not bind users to a specific encoding — encoding and the SNN itself are orthogonal. The framework just provides encoders that you can plug in front of the network.

Built-in encoders in `spikingjelly.activation_based.encoding`:

- `PoissonEncoder` — Poisson / rate coding (stateless)
- `LatencyEncoder` — latency / TTFS coding (stateful)
- `PeriodicEncoder` — periodic spike train
- `WeightedPhaseEncoder` — weighted phase
- `PopSpikeEncoderDeterministic` / `PopSpikeEncoderRandom` — population coding
- `PopEncoder` — population rate encoding

Quick guidelines:
- For general performance, prefer **rate coding (Poisson)**.
- For sparsity / energy, consider **latency / TTFS** (harder to train).
- For precise spike-timing control, use **TTFS** with a custom neuron.

Tutorial: https://spikingjelly.readthedocs.io/en/latest/activation_based/4_encoding.html
